agent/consul/state/event_types.go

```diff
@@ -0,0 +1,107 @@
+package state
+
+import (
+	"context"
+
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/structs"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+// ServiceHealthSnapshot returns stream.Events that provide a snapshot of the
+// current state.
+func (s *Store) ServiceHealthSnapshot(ctx context.Context, eventCh chan stream.Event, service string) error {
+	idx, nodes, err := s.CheckServiceNodes(nil, service)
+	if err != nil {
+		return err
+	}
+
+	checkServiceNodesToServiceHealth(idx, nodes, ctx, eventCh)
+
+	return nil
+}
+
+// RegistrationEvents returns stream.Events that correspond to a catalog
+// register operation.
+func (s *Store) RegistrationEvents(tx *memdb.Txn, node, service string) ([]stream.Event, error) {
+	idx := maxIndexTxn(tx, "nodes", "services")
+
+	_, services, err := s.nodeServicesTxn(tx, nil, node, service, false)
+	if err != nil {
+		return nil, err
+	}
+
+	idx, nodes, err := s.parseCheckServiceNodes(tx, nil, idx, "", services, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return checkServiceNodesToServiceHealth(idx, nodes, nil, nil), nil
+}
+
+// checkServiceNodesToServiceHealth converts a list of CheckServiceNodes to
+// ServiceHealth events for streaming. If a non-nil channel and context are passed,
+// the events will be sent to the channel instead of appended to a slice.
+func checkServiceNodesToServiceHealth(idx uint64, nodes structs.CheckServiceNodes,
```

Comment:
This is IMHO bad design to have a func that behaves differently based on whether a param is `nil`. A caller woudnt expect the returned slice to be nil if a chan was passed.

Reply:
I agree it's not the nicest but it is internal only and has the benefit of not needing to duplicate the logic here between two different delivery mechanisms.

We can't just make the streaming/chan version on top of the slice one because then we'd have to create the the entire slice in memory first before sending anything which is what we want to avoid in streaming. We could make this be streaming first and then build the non-streaming version on top by iterating the chan and turning it into a slice but that seems wasteful in that common case to do all the extra chan allocations for nothing.

I guess we could abstract and have a `func(e stream.Event) error` argument that can be passed a handler that either appends to a slice or sends on a chan depending on the need (and stops iteration on error). Do you think that's cleaner @Aestek ?